File: dungeons-26.1.2/src/main/java/fishmod/utils/data/ItemUtil.java
package fishmod.utils.data;

import java.util.List;
import net.minecraft.client.Minecraft;
import net.minecraft.client.player.LocalPlayer;
import net.minecraft.core.component.DataComponents;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.chat.Component;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.component.CustomData;
import net.minecraft.world.item.component.ItemLore;

public class ItemUtil {

    public static String getId(ItemStack item) {
        CustomData nbt = item.get(DataComponents.CUSTOM_DATA);
        if (nbt == null) return null;

        CompoundTag compound = nbt.copyTag();

        return compound.getStringOr("id", null);
    }

    /** Reads a raw string value out of an item's CUSTOM_DATA NBT (e.g. "petInfo"), or null. */
    public static String getNbtString(ItemStack item, String key) {
        CustomData nbt = item.get(DataComponents.CUSTOM_DATA);
        if (nbt == null) return null;
        return nbt.copyTag().getStringOr(key, null);
    }

    public static String getUuid(ItemStack item) {
        CustomData nbt = item.get(DataComponents.CUSTOM_DATA);
        if (nbt == null) return null;

        CompoundTag compound = nbt.copyTag();

        return compound.getStringOr("uuid", null);
    }

    public static boolean isHolding(String name) {
        LocalPlayer player = Minecraft.getInstance().player;
        if (player == null) return false;
        return player.getMainHandItem().getHoverName().getString().contains(name);
    }

    public static boolean itemHasName(ItemStack itemStack, String name) {
        LocalPlayer player = Minecraft.getInstance().player;
        if (player == null || itemStack == null) return false;
        return itemStack.getHoverName().getString().contains(name);
    }

    public static boolean containsLore(ItemStack item, String contain) {
        if (item ==null) return false;
        ItemLore lore = item.get(DataComponents.LORE);
        if (lore == null) return false;

        List<Component> lines = lore.lines();
        if (lines.isEmpty()) return false;

        for (Component line : lines.reversed()) {
            String string = line.getString();
            if (string.contains(contain)) {
                return true;
            }
        }
        return false;
    }

    public static boolean containsIgnoreCaseLore(ItemStack item, String contain) {
        if (item ==null) return false;
        ItemLore lore = item.get(DataComponents.LORE);
        if (lore == null) return false;

        List<Component> lines = lore.lines();
        if (lines.isEmpty()) return false;

        for (Component line : lines.reversed()) {
            String string = line.getString();
            if (string.toLowerCase().contains(contain.toLowerCase())) {
                return true;
            }
        }
        return false;
    }

    public static boolean containsNBT(ItemStack item, String contain) {
        CustomData nbt = item.get(DataComponents.CUSTOM_DATA);
        if (nbt == null) {
            return false;
        }

        return  nbt.toString().contains(contain);
    }

}

File: dungeons-26.1.2/src/main/kotlin/fishmod/utils/data/ItemUtil.kt
package fishmod.utils.data

import fishmod.features.item.fishmodCustomDataTag
import net.minecraft.core.component.DataComponents
import net.minecraft.world.item.ItemStack

object ItemUtil {

    @JvmStatic
    fun getId(item: ItemStack): String? {
        val compound = item.fishmodCustomDataTag() ?: return null
        return if (compound.contains("id")) compound.getStringOr("id", "") else null
    }

    @JvmStatic
    fun getUuid(item: ItemStack): String? {
        val compound = item.fishmodCustomDataTag() ?: return null
        return if (compound.contains("uuid")) compound.getStringOr("uuid", "") else null
    }

    @JvmStatic
    fun containsLore(item: ItemStack?, contain: String): Boolean {
        if (item == null) return false
        val lore = item.get(DataComponents.LORE) ?: return false

        val lines = lore.lines()
        if (lines.isEmpty()) return false

        for (line in lines.asReversed()) {
            val string = line.string
            if (string.contains(contain)) {
                return true
            }
        }
        return false
    }

    @JvmStatic
    fun containsIgnoreCaseLore(item: ItemStack?, contain: String): Boolean {
        if (item == null) return false
        val lore = item.get(DataComponents.LORE) ?: return false

        for (line in lore.lines().asReversed()) {
            if (line.string.contains(contain, ignoreCase = true)) return true
        }
        return false
    }
}
